### PR TITLE
Feature spire

### PR DESCRIPTION
**Pull request by pxp928**
**Fixes issue:**

**Description:**
Adding Spire integration to In-toto-golang. Thanks to @mikhailswift for helping update the spire implementation. Added new unit testing, dockerfile and docker-compose to bring up test infrastructure with spire server and spire agent for testing. Updates made to the readme and Makefile to allow for easy testing of the spire components. 

**Please verify and check that the pull request fulfills the following
requirements:**

- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+
+RUN apk --no-cache add make ca-certificates 
+WORKDIR /src
+COPY go.mod go.sum /src/
+RUN go mod download
+COPY . /src/
+RUN make build
+ENTRYPOINT [ "/src/bin/in-toto" ]
+
+FROM debian:buster
```
Comment (shibumi) [inline Dockerfile:13]: I suggest we just use Google's distroless containers here. They have a smaller attack surface and provide ca certificates as well :)

`FROM gcr.io/distroless/base`

File: Dockerfile [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+FROM golang:${GO_VERSION}-alpine as build
+
+RUN apk --no-cache add make ca-certificates 
+WORKDIR /src
+COPY go.mod go.sum /src/
+RUN go mod download
+COPY . /src/
+RUN make build
+ENTRYPOINT [ "/src/bin/in-toto" ]
```
Comment (shibumi) [inline Dockerfile:11]: I think you can remove  the ENTRYPOINT here.. it's just a build layer.

File: Dockerfile [hunk 3]
```diff
@@ -0,0 +1,16 @@
+ARG GO_VERSION=1.17
+
+FROM golang:${GO_VERSION}-alpine as build
+
+RUN apk --no-cache add make ca-certificates
```
Comment (shibumi) [inline Dockerfile:5]: are the ca-certificates really necessary for the build?

File: cmd/record.go [hunk 4]
```diff
@@ -111,6 +110,22 @@ of another.`,
...
 	)
 
+	recordCmd.PersistentFlags().StringVar(
+		&spiffeUDS,
+		"spiffe-workload-api-path",
+		"",
+		"UDS path for SPIFFE workload API",
+	)
+
+	recordCmd.PersistentFlags().BoolVar(
```
[the remark this answers is not in the document]
Comment (shibumi) [reply cmd/record.go:120]: makes sense for me!  :+1:

File: doc/in-toto.md [hunk 5]
```diff
@@ -23,3 +23,4 @@ A framework to secure the integrity of software supply chains https://in-toto.io
 * [in-toto sign](in-toto_sign.md)	 - Provides command line interface to sign in-toto link or layout metadata
 * [in-toto verify](in-toto_verify.md)	 - Verify that the software supply chain of the delivered product
 
+###### Auto generated by spf13/cobra on 18-Oct-2021
```
Comment (shibumi) [inline doc/in-toto.md:26]: Can you remove these comments from the PR via setting  DisableAutoGenTag: true, in the cobra RootCmd again?

File: cmd/root.go [hunk 6]
```diff
@@ -21,12 +26,86 @@ var (
...
-	Short:             "Framework to secure integrity of software supply chains",
-	Long:              `A framework to secure the integrity of software supply chains https://in-toto.io/`,
-	SilenceUsage:      true,
-	SilenceErrors:     true,
-	DisableAutoGenTag: true,
+	Use:           "in-toto",
+	Short:         "Framework to secure integrity of software supply chains",
+	Long:          `A framework to secure the integrity of software supply chains https://in-toto.io/`,
+	SilenceUsage:  true,
+	SilenceErrors: true,
```
Comment (shibumi) [inline cmd/root.go:33]: Please set DisableAutoGenTag: true again.

File: internal/spiffe/spiffe.go [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+	"crypto"
+	"crypto/x509"
+	"encoding/pem"
+	"fmt"
+
+	intoto "github.com/in-toto/in-toto-golang/in_toto"
+	"github.com/spiffe/go-spiffe/v2/workloadapi"
+)
+
+type SVIDDetails struct {
```
Comment (shibumi) [inline internal/spiffe/spiffe.go:15]: Would be cool to have some docstrings here. What do you think? :)

File: internal/spiffe/spiffe.go [hunk 8]
```diff
@@ -0,0 +1,77 @@
...
+	PrivateKey    crypto.Signer
+	Certificate   *x509.Certificate
+	Intermediates []*x509.Certificate
+}
+
+type SVIDFetcher interface {
```
Comment (shibumi) [inline internal/spiffe/spiffe.go:21]: Would be cool to have some docstrings here. What do you think? :)

File: internal/spiffe/spiffe.go [hunk 9]
```diff
@@ -0,0 +1,77 @@
...
+		return s, fmt.Errorf("svid has no key")
+	}
+
+	s.PrivateKey = svid.PrivateKey
+	s.Certificate = svid.Certificates[0]
+	s.Intermediates = svid.Certificates[1:]
+	return s, nil
+}
+
+func (s SVIDDetails) InTotoKey() (intoto.Key, error) {
```
Comment (shibumi) [inline internal/spiffe/spiffe.go:58]: Would be cool to have some docstrings here. What do you think? :)

File: test-infra/infra-down.sh [hunk 10]
```diff
@@ -0,0 +1,5 @@
+#! /bin/sh
```
Comment (shibumi) [inline test-infra/infra-down.sh:1]: Is there a whitespace between #! and /bin/sh? Shouldn't this be `#!/bin/sh` or is this still a valid shebang line? oO

File: test-infra/infra-up.sh [hunk 11]
```diff
@@ -0,0 +1,5 @@
+#! /bin/sh
```
Comment (shibumi) [inline test-infra/infra-up.sh:1]: Is there a whitespace between #! and /bin/sh? Shouldn't this be #!/bin/sh or is this still a valid shebang line? oO

File: test-infra/register.sh [hunk 12]
```diff
@@ -0,0 +1,20 @@
+#!/bin/bash
+
+## DEMO REGISTRATIONS #######################################
```
Comment (shibumi) [inline test-infra/register.sh:3]: Do we really need the header and footer comment in this file?

I mean: ## DEMO REGISTRATIONS and ... ## ------

File: Dockerfile (hunk 1 above)
Comment (pxp928) [reply Dockerfile]: I agree that is better. Changed.

File: Dockerfile (hunk 2 above)
Comment (pxp928) [reply Dockerfile]: Removed EntryPoint as its not needed.

File: Dockerfile (hunk 3 above)
Comment (pxp928) [reply Dockerfile]: ca-certificate was not needed. Left over from a previous test. Removed.

File: doc/in-toto.md (hunk 5 above)
Comment (pxp928) [reply doc/in-toto.md]: Added DisableAutoGenTag: true to remove comments.

File: internal/spiffe/spiffe.go (hunk 7 above)
Comment (pxp928) [reply internal/spiffe/spiffe.go:19]: Added missing docstring.

File: cmd/root.go (hunk 6 above)
Comment (pxp928) [reply cmd/root.go]: Added DisableAutoGenTag: true to remove comments.

File: internal/spiffe/spiffe.go (hunk 8 above)
Comment (pxp928) [reply internal/spiffe/spiffe.go:29]: Added missing docstring.

File: internal/spiffe/spiffe.go (hunk 9 above)
Comment (pxp928) [reply internal/spiffe/spiffe.go:74]: Added missing docstring.

File: test-infra/infra-down.sh (hunk 10 above)
Comment (pxp928) [reply test-infra/infra-down.sh]: removed whitespace. Strange it did not error out.

File: test-infra/infra-up.sh (hunk 11 above)
Comment (pxp928) [reply test-infra/infra-up.sh]: removed whitespace. Strange it did not error out.

File: test-infra/register.sh (hunk 12 above)
Comment (pxp928) [reply test-infra/register.sh]: Removed header/footer comment

**Review by shibumi — APPROVED** [review]

**Comment by adityasaky** [comment]
Thank you for all the fantastic work here, @mikhailswift!

**Pull request merged**